ArcGISRuntimeSDKQt_CppSamples/DisplayInformation/AddGraphicsWithRenderer/main.cpp

```diff
@@ -11,76 +11,80 @@
-
-  // Use of Esri location services, including basemaps and geocoding,
-  // requires authentication using either an ArcGIS identity or an API Key.
-  // 1. ArcGIS identity: An ArcGIS named user account that is a member of an
-  //    organization in ArcGIS Online or ArcGIS Enterprise.
-  // 2. API key: A permanent key that gives your application access to Esri
-  //    location services. Visit your ArcGIS Developers Dashboard create a new
-  //    API keys or access an existing API key.
+  app.setApplicationName(QString("AddGraphicsWithRenderer - C++"));
+
+  // Access to Esri location services requires an API key. This can be copied below or used as a command line argument.
   const QString apiKey = QString("");
-  if (apiKey.isEmpty())
-  {
-      qWarning() << "Use of Esri location services, including basemaps, requires" <<
-                    "you to authenticate with an ArcGIS identity or set the API Key property.";
-  }
-  else
-  {
-      Esri::ArcGISRuntime::ArcGISRuntimeEnvironment::setApiKey(apiKey);
-  }
+  setAPIKey(app, apiKey);
+
   // Initialize the sample
   AddGraphicsWithRenderer::init();
 
   // Initialize application view
-  QQuickView view;
-  view.setResizeMode(QQuickView::SizeRootObjectToView);
-
+  QQmlApplicationEngine engine;
   // Add the import Path
-  view.engine()->addImportPath(QDir(QCoreApplication::applicationDirPath()).filePath("qml"));
-
-  QString arcGISRuntimeImportPath = QUOTE(ARCGIS_RUNTIME_IMPORT_PATH);
+  engine.addImportPath(QDir(QCoreApplication::applicationDirPath()).filePath("qml"));
 
- #if defined(LINUX_PLATFORM_REPLACEMENT)
-  // on some linux platforms the string 'linux' is replaced with 1
-  // fix the replacement paths which were created
-  QString replaceString = QUOTE(LINUX_PLATFORM_REPLACEMENT);
-  arcGISRuntimeImportPath = arcGISRuntimeImportPath.replace(replaceString, "linux", Qt::CaseSensitive);
- #endif
-
-  // Add the Runtime and Extras path
-  view.engine()->addImportPath(arcGISRuntimeImportPath);
+#ifdef ARCGIS_RUNTIME_IMPORT_PATH_2
+  engine.addImportPath(ARCGIS_RUNTIME_IMPORT_PATH_2);
+#endif
 
   // Set the source
-  view.setSource(QUrl("qrc:/Samples/DisplayInformation/AddGraphicsWithRenderer/AddGraphicsWithRenderer.qml"));
-
-  view.show();
+  engine.load(QUrl("qrc:/Samples/DisplayInformation/AddGraphicsWithRenderer/main.qml"));
 
   return app.exec();
 }
 
+// Use of Esri location services, including basemaps and geocoding,
+// requires authentication using either an ArcGIS identity or an API Key.
+// 1. ArcGIS identity: An ArcGIS named user account that is a member of an
+//    organization in ArcGIS Online or ArcGIS Enterprise.
+// 2. API key: API key: a permanent key that grants access to
+//    location services and premium content in your applications.
+//    Visit your ArcGIS Developers Dashboard to create a new
+//    API key or access an existing API key.
+
+void setAPIKey(const QGuiApplication& app, QString apiKey)
```

Comment:
I'm not sure I understand. There's a method in the sample viewer that sets the API key because we don't use individual samples' main.cpp. Do you mean you'd like a way for us to set an API key in one common place that all samples can reference? Because I'd be all for that -- placing the API key there manually every time I open a new sample gets annoying 😅 . I'd really like if it could read an API key from my system's environment variables, but that's just one of a few solutions.

Reply:
I am talking purely about this logic and not about how we pass or get the key from the build. This same chunk of code is duplicated in numerous samples I saw, and reducing that duplication is my goal. I know we need to make code standards concessions to keep the samples usable as independent pieces, so that suggestions may go against that.

Reply:
Woops, I meant to link the method - https://github.com/Esri/arcgis-runtime-samples-qt/blob/v.next/ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_Samples/SampleManager.cpp#L348

Reply:
Right, unfortunately, pretty much the entire main.cpp apart from the included sample header file and the sample app name is duplicated for each sample, that method included. I think this one is just going to be a consequence of being able to build samples individually... unless we want to link a common main.cpp at the root directory? Hm...